linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -2100,141 +2017,123 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
     fb.buildAndAddToModule()
   }
 
-  /** `checkCast: (ref typeData), anyref -> []`.
+  /** `cast: (ref jlClass), anyref -> anyref`.
    *
    *  Casts the given value to the given type; subject to undefined behaviors.
+   *
+   *  This is the underlying func for the `Class_cast` operation.
    */
-  private def genCheckCast()(implicit ctx: WasmContext): Unit = {
-    val typeDataType = RefType(genTypeID.typeData)
+  private def genCast()(implicit ctx: WasmContext): Unit = {
+    assert(semantics.asInstanceOfs != CheckedBehavior.Unchecked)
 
-    val fb = newFunctionBuilder(genFunctionID.checkCast)
-    val typeDataParam = fb.addParam("typeData", typeDataType)
+    val fb = newFunctionBuilder(genFunctionID.cast)
+    val jlClassParam = fb.addParam("jlClass", RefType(genTypeID.ClassStruct))
     val valueParam = fb.addParam("value", RefType.anyref)
+    fb.setResultType(RefType.anyref)
 
-    if (semantics.asInstanceOfs != CheckedBehavior.Unchecked) {
-      fb.block() { successLabel =>
-        // If typeData.kind == KindJSType, succeed
-        fb += LocalGet(typeDataParam)
-        fb += StructGet(genTypeID.typeData, genFieldID.typeData.kind)
-        fb += I32Const(KindJSType)
-        fb += I32Eq
-        fb += BrIf(successLabel)
-
-        // If value is null, succeed
-        fb += LocalGet(valueParam)
-        fb += RefIsNull // consumes `value`, unlike `BrOnNull` which would leave it on the stack
-        fb += BrIf(successLabel)
+    val typeDataLocal = fb.addLocal("typeData", RefType(genTypeID.typeData))
 
-        // If isInstance(typeData, value), succeed
-        fb += LocalGet(typeDataParam)
-        fb += LocalGet(valueParam)
-        fb += Call(genFunctionID.isInstance)
-        fb += BrIf(successLabel)
-
-        // Otherwise, it is a CCE
-        fb += LocalGet(valueParam)
-        fb += LocalGet(typeDataParam)
-        fb += Call(genFunctionID.classCastException)
-        fb += Unreachable // for clarity; technically redundant since the stacks align
-      }
-    }
+    fb.block() { successLabel =>
+      // typeData := jlClass.data, leave it on the stack for the kind test
+      fb += LocalGet(jlClassParam)
+      fb += StructGet(genTypeID.ClassStruct, genFieldID.classData)
+      fb += LocalTee(typeDataLocal)
 
-    fb.buildAndAddToModule()
-  }
+      // If typeData.kind == KindJSType, succeed
+      fb += StructGet(genTypeID.typeData, genFieldID.typeData.kind)
+      fb += I32Const(KindJSType)
+      fb += I32Eq
+      fb += BrIf(successLabel)
 
-  /** `getComponentType: (ref typeData) -> (ref null jlClass)`.
-   *
-   *  This is the underlying func for the `getComponentType()` closure inside class data objects.
```

Comment:
Tying to my comment above: I think we should keep this.

Reply:
Huh, yes, indeed. That's much better.